docs/rfcs/sync-manager.md

```diff
@@ -0,0 +1,171 @@
+
+The primary responsibilities of the engine are:
+
+* Manage the "meta/global" resource. This resources holds the following information:
+    * syncID - the global syncID for the entire storage managed by sync.
+    * storageVersion - an integer. If this value is greater than a sync client knows about, that client will refuse to sync. While this is rarely changed, it allows us to lock out older sync clients which could be useful.
+    * The list of "declined" engines - that is, engines which the user has explicitly declined from syncing. Note that declined engines are global for the account - declining an engine on any device should decline it on every device. While this is something we'd like to consider changing in the future, for now we are sticking with this global declined state.
+    * Some metadata about every collection (which may be subtly different from an "engine"). Currently this metadata is only {version, syncID}
+
+    Management of meta/global includes all syncIDs - if there's no global syncID, or an engine has no syncID, it should generate one. As we mention below, engines will be responsible for persisting the syncID and version and taking necessary action when it changes. Thus, engines never need to communicate a new syncID back to the global.
+
+* Query info/collections and pass this information to engines.
+
+* Manage interaction with the token server to obtain a token. As part of this, the manager needs to detect when an engine fails to use the token, generally due to a node reassignment mid sync. In this scenario, the manager will abort any pending syncs and reset itself, starting a new sync flow. Note also that this means the containing application will generally not need to see authentication errors from engines, only from the manager.
+
+* Manage the "clients" collection - we probably can't ignore this any longer, especially for bookmarks (as desktop will send a wipe command on bookmark restore). This involves both communicating with the engine regarding commands targetting it, and accepting commands to be send to other devices. Note however that it's not yet clear where these commands will originate from - eg, consider a bookmark restore - which point in this process will queue the wipe commands for other clients?). Exactly how to manage this (eg, when we can assume the engine has processed the command?) are TBD.
+
+* Perform, or coordinate, the actual sync of the engines - from the containing app's POV, there's a single "sync now" entry-point. Exactly how the manager binds to engines is TBD. For example, a possible implementation is that a callback is registered with the sync manager. It's also likely that we will expose a way to sync a specific engine rather than "sync all".
```

Comment:
> For example, a possible implementation is that a callback is registered with the sync manager.

I'm unsure how this can be done safely. To be safe the callback would need to be `+ 'static`, which severely limits what we can do. (And that's assuming we're megazorded, it would be extremely unsafe outside of megazording)

I think realistically it will probably have to return a list of engines (e.g. probably strings or enums) over the FFI.

Reply:
One thing to consider is shipping the monolith first: for MVP, require Megazording, bundle all engines, and sync all, not a specific engine. This means Lockbox and other consumers will temporarily pull in code for bookmarks and history, even if they don't use them—though we could maybe disable storage for those engines, and our Sync adapter could detect that and skip syncing the engine?

Once we have the Sync manager in place, we could go back, see where the boundaries are, and figure out how to split them up. I think I'm a bit wary of trying to figure out how to modularize this now, especially when the FFI is involved, when we haven't built it.

Reply:
> One thing to consider is shipping the monolith first: for MVP, require Megazording, bundle all engines, and sync all, not a specific engine.

I agree (and should have been more explicit about this). In the first cut, I expect that if the rust sync manager is used, all engines will be rust implemented - this will be true for Fenix and Lockbox. iOS (and later, desktop) is a special case where the manager is likely to not be rust implemented, but that seems relatively easy too - it can just make regular ffi calls into the rust engines.

Reply:
I removed all references to callbacks and added some extra comments around this.